packages/core/graph/src/AdjacencyList.js

```diff
@@ -0,0 +1,1195 @@
+      (Array.isArray(type) && type.includes(ALL_EDGE_TYPES));
+
+    let nodes = [];
+    if (isAllEdgeTypes) {
+      for (let toSet of typeMap.getConnectedFrom(from).values()) {
+        nodes.push(...toSet);
+      }
+    } else if (Array.isArray(type)) {
+      for (let typeNum of type) {
+        nodes.push(...typeMap.getConnectedFromType(from, typeNum));
+      }
+    } else {
+      nodes.push(...typeMap.getConnectedFromType(from, (type: any)));
+    }
+    return nodes;
+  }
+
+  /**
+   * Get the list of nodes connected to this node.
+   */
+  getNodesConnectedTo(
+    to: NodeId,
+    type:
+      | AllEdgeTypes
+      | TEdgeType
+      | NullEdgeType
+      | Array<TEdgeType | NullEdgeType> = 1,
+  ): NodeId[] {
+    let typeMap = this.#typeMap || (this.#typeMap = new TypeMap(this));
+    let isAllEdgeTypes =
+      type === ALL_EDGE_TYPES ||
+      (Array.isArray(type) && type.includes(ALL_EDGE_TYPES));
+
+    let nodes = [];
+    if (isAllEdgeTypes) {
+      for (let fromSet of typeMap.getConnectedTo(to).values()) {
+        nodes.push(...fromSet);
+      }
+    } else if (Array.isArray(type)) {
+      for (let typeNum of type) {
+        nodes.push(...typeMap.getConnectedToType(to, typeNum));
+      }
+    } else {
+      nodes.push(...typeMap.getConnectedToType(to, (type: any)));
+    }
+    return nodes;
+  }
+
+  /**
+   * Create a hash of the edge connecting the `from` and `to` nodes.
+   *
+   * This hash is used to index the edge in the `edges` array.
+   *
+   */
+  hash(from: NodeId, to: NodeId, type: TEdgeType | NullEdgeType): EdgeHash {
+    // A crude multiplicative hash, in 3 steps:
+    // 1. Serialize the args into an integer that reflects the argument order,
+    // shifting the magnitude of each argument by the sum
+    // of the significant digits of the following arguments,
+    // .e.g., `hash(10, 24, 4) => 10244`.
+    // $FlowFixMe[unsafe-addition]
+    // $FlowFixMe[incompatible-type]
+    let hash = '' + from + to + type - 0;
+    // 2. Mix the upper bits of the integer into the lower bits.
+    // We do this to increase the likelihood that a change to any
+    // bit of the input will vary the output widely.
+    hash = hash32shift(hash);
+    // 3. Map the hash to a value modulo the edge capacity.
+    hash %= this.#edges[CAPACITY];
+    return hash;
+  }
+}
+
+/**
+ * A cache of connected nodes grouped by type.
+ *
+ * This cache is used to speed up iterations that are grouped by type,
+ * such as `getEdgesConnectedFrom` and `getEdgesConnectedTo`,
+ * as well as `getOutboundEdgesByType` and `getInboundEdgesByType`.
+ */
+class TypeMap<TEdgeType: number = 1> {
+  #data: AdjacencyList<TEdgeType>;
+  /** A map of node ids from => through types => to node ids. */
+  #from: Map<NodeId, Map<TEdgeType | NullEdgeType, Set<NodeId>>> = new Map();
+  /** A map of node ids to => through types => from node ids. */
+  #to: Map<NodeId, Map<TEdgeType | NullEdgeType, Set<NodeId>>> = new Map();
```

Comment:
How much do these extra maps affect performance? We're basically paying the cost of 3 hash maps instead of just one, so basically the cost of our old implementation PLUS the cost of your new implementation (minus serializing these maps). Explains why it is slightly slower in some cases.

Would it make sense to adjust the main adjacency list data structure instead of be more oriented around iterating by type? It's pretty rare that we access all of the edges connected to a node across all edge types. Most of the time, we are asking for a single type. One idea could be to change the NEXT_IN, NEXT_OUT, etc. pointers to be the next edge of the same type rather than having a single linked list of every edge connected to a node. The nodes would gain a type in addition to FIRST_IN/FIRST_OUT etc., so basically there'd be a "node" per id + type. This way you'd basically have a separate linked list per edge type and could simply walk that to iterate the connected edges. If you wanted all edge types, you could instead pass a list of those (which we know upfront).

Reply:
We refactored the data structure just as you suggested; These maps are gone now! 🎉